service/controller/v15/adapter/spec.go

```diff
@@ -89,13 +89,19 @@ type StackState struct {
 	MasterImageID              string
 	MasterInstanceType         string
 	MasterInstanceResourceName string
-	MasterCloudConfigVersion   string
-	MasterInstanceMonitoring   bool
+	// TODO the cloud config versions shouldn't be injected here. These should
+	// actually always only be the ones the operator has hard coded. No other
+	// version should be used here ever.
```

Comment:
I don't follow this TODO. Where are the cc versions hardcoded? Do you mean the version in the key package?

Reply:
Yes.

Reply:
Sorry I'm being slow here. I agree the source should be the key package. But why is it no longer needed in the adapter? Is it because we only use the S3 URLs in the templating?

Reply:
The computation of the S3 URLs was formerly done like everywhere. This is now in one place, the key package. The key package already holds the CC version and it actually never changes. I do not understand why it was configurable in the first place because it should actually only ever be what we have hard coded.

Reply:
Thanks for the explanation I get it now.

I agree we should compute the S3 URLs once and use them everywhere.  I just missed that when I read the comment.